morph/react/block-list.js

```diff
@@ -49,21 +50,45 @@ export function enter(node, parent, state) {
   }
 
   state.render.push(
-    `{Array.isArray(${value}) && ${value}.map((item, index, list) => `
+    `{Array.isArray(${value}) && ${value}.map((item, index, list) => {`
   )
 
   if (state.viewPath) {
     let itemDataContextName =
       getProp(node, 'itemDataContextName') ||
       defaultItemDataContextName(node, from)
+    let keys
+    if (hasProp(node, 'itemKeys')) {
+      keys = getProp(node, 'itemKeys')
+        .value.split(',')
+        .map((key) => key.trim())
+    } else {
+      keys = ['id']
+
+      state.render.push(`
+        if (process.env.NODE_ENV === 'development') {
+          if(!item?.id) {
+            console.debug({
+              type: 'views/data',
+              warning: \`Missing "id" property on the item $\{JSON.stringify(item)} in context "${itemDataContextName.value}" so the index in the list will be used as a fallback key. Consider setting "itemKeys" if the item has a custom or compound key e.g. itemKeys first_name,last_name.\`,
+            })
+          }
+        }`)
+    }
+    let computedKey =
+      keys.length === 1
+        ? `item?.${keys[0]} || index`
+        : `${JSON.stringify(
+            keys
+          )}.map(key => item?.[key]).filter(Boolean).join('-') || index`
```

Comment:
Could we expand the code to be the final version of it so it reads
```
`${item?.key1}-${item?.key2}`
```

I know that your code would remove falsy values but it also needs to iterate over an array every time it gets rendered. Alternatively, the key's name could be used as a placeholder:

```
`${item?.key1 || "key1"}-${item?.key2 || "key2"}`
```

Reply:
Thinking a bit more about it, maybe it doesn't make much sense to have that falsy filter, if the developer provides a key then it should probably have a value. Do you think it makes sense to remove it and will key something that reads like the first example you suggested?

Reply:
Yeah, I think it makes sense not to have a map/filter and just have the key in place. I'd also go to the extent of not using a default because it's the developer's responsibility to have the values in place as you well mentioned too.